client-rpc/src/rpc/websocket_core.rs

```diff
@@ -0,0 +1,376 @@
+
+            max_height: 0,
+            state: WebsocketState::ReadyProcess,
+            global_state_service: gss,
+            client,
+            block_handler,
+            wallets,
+            current_wallet: 0,
+
+            wallet_client,
+        }
+    }
+
+    // to process multiple wallets
+    fn get_current_wallet(&self) -> WalletInfo {
+        assert!(self.current_wallet < self.wallets.len());
+        self.wallets[self.current_wallet].clone()
+    }
+
+    // get height from database
+    fn get_current_height(&self) -> u64 {
+        let wallet = self.get_current_wallet();
+        self.global_state_service
+            .last_block_height(&wallet.view_key)
+            .unwrap()
+    }
+
+    // write block to internal database
+    fn do_save_block_to_chain(&mut self, block: Block, kind: &str) {
+        if self.wallets.is_empty() {
+            return;
+        }
+        // restore as object
+        let height: u64 = block.block.header.height.parse().unwrap();
+        let current = self.get_current_height();
+        if height != current + 1 {
+            log::info!(
+                "drop block {} current={} max={}",
+                height,
+                current,
+                self.max_height
+            );
+            return;
+        }
+        log::info!("save block height={} kind={}", height, kind);
+        let _ = self.write_block(height, &block);
+    }
+
+    // low level block processing
+    pub fn write_block(&self, block_height: u64, block: &Block) -> Result<()> {
+        let block_results = self.client.block_results(block_height)?;
+
+        let block_time = block.time();
+
+        let transaction_ids = block_results.transaction_ids()?;
+        let block_filter = block_results.block_filter()?;
+
+        let wallet = self.get_current_wallet();
+        let unencrypted_transactions = self.check_unencrypted_transactions(
+            &block_filter,
+            wallet.staking_addresses.as_slice(),
+            block,
+        )?;
+
+        let block_header = BlockHeader {
+            block_height,
+            block_time,
+            transaction_ids,
+            block_filter,
+            unencrypted_transactions,
+        };
+
+        self.block_handler
+            .on_next(block_header, &wallet.view_key, &wallet.private_key)?;
+        Ok(())
+    }
+
+    // now one session is complete
+    // let's wait some time to next blocks
+    pub fn change_to_wait(&mut self) {
+        self.state = WebsocketState::WaitProcess;
+        self.state_time = SystemTime::now();
+    }
+
+    // tx channel for this thread
+    pub fn get_queue(&self) -> Sender<OwnedMessage> {
+        self.my_sender.clone()
+    }
+
+    // because everything is done via channel
+    // no mutex is necessary
+    // wallet can be added in runtime
+    pub fn add_wallet(&mut self, name: String, passphrase: SecUtf8) -> JsonResult<()> {
+        log::info!("add_wallet ***** {} {}", name, passphrase);
```

Comment:
I think passphrase should not be printed here? Although it is protected by SecUtf8 already

Reply:
ok